crates/stages/src/stages/index_storage_history.rs

```diff
@@ -57,16 +78,61 @@ impl<DB: Database> Stage<DB> for IndexStorageHistoryStage {
         let (range, unwind_progress, is_final_range) =
             input.unwind_block_range_with_threshold(self.commit_threshold);
 
-        tx.unwind_storage_history_indices(BlockNumberAddress::range(range))?;
+        let changesets = tx.unwind_storage_history_indices(BlockNumberAddress::range(range))?;
+
+        let checkpoint =
+            if let Some(mut stage_checkpoint) = input.checkpoint.index_history_stage_checkpoint() {
+                stage_checkpoint.progress.processed -= changesets as u64;
+                StageCheckpoint::new(unwind_progress)
+                    .with_index_history_stage_checkpoint(stage_checkpoint)
+            } else {
+                StageCheckpoint::new(unwind_progress)
+            };
 
         info!(target: "sync::stages::index_storage_history", to_block = input.unwind_to, unwind_progress, is_final_range, "Unwind iteration finished");
-        Ok(UnwindOutput { checkpoint: StageCheckpoint::new(unwind_progress) })
+        Ok(UnwindOutput { checkpoint })
     }
 }
 
+fn stage_checkpoint<DB: Database>(
```

Comment:
Would be ideal if we could be generic over `T: Table` here instead of duplicating the function but the `.walk_range(BlockNumberAddress::range(0..=checkpoint.block_number))?` makes it hard I think.